log4j-api/src/main/java/org/apache/logging/log4j/spi/LoggingSystem.java

```diff
@@ -0,0 +1,363 @@
+import java.util.SortedMap;
+import java.util.TreeMap;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+/**
+ * Handles initializing the Log4j API through {@link Provider} discovery. This keeps track of which
+ * {@link LoggerContextFactory} to use in {@link LogManager} along with factories for {@link ThreadContextMap}
+ * and {@link ThreadContextStack} to use in {@link ThreadContext}.
+ *
+ * @since 3.0.0
+ */
+public class LoggingSystem {
+    /**
+     * Resource name for a Log4j 2 provider properties file.
+     */
+    private static final String PROVIDER_RESOURCE = "META-INF/log4j-provider.properties";
+    private static final String API_VERSION = "Log4jAPIVersion";
+    private static final String[] COMPATIBLE_API_VERSIONS = {"3.0.0"};
+
+    public static final String THREAD_CONTEXT_MAP_DISABLED = "log4j2.disableThreadContextMap";
+    public static final String THREAD_CONTEXT_STACK_DISABLED = "log4j2.disableThreadContextStack";
+    public static final String THREAD_CONTEXT_DISABLED = "log4j2.disableThreadContext";
+    /**
+     * Property name ({@value} ) for selecting {@code InheritableThreadLocal} (value "true") or plain
+     * {@code ThreadLocal} (value is not "true") in the {@link ThreadContextMap} implementation.
+     */
+    public static final String THREAD_CONTEXT_MAP_INHERITABLE_ENABLED = "log4j2.isThreadContextMapInheritable";
+    public static final String THREAD_CONTEXT_MAP_CLASS_NAME = "log4j2.threadContextMap";
+    public static final String THREAD_CONTEXT_INITIAL_CAPACITY = "log4j2.threadContextInitialCapacity";
+    public static final int THREAD_CONTEXT_DEFAULT_INITIAL_CAPACITY = 16;
+    public static final String THREAD_CONTEXT_GARBAGE_FREE_ENABLED = "log4j2.garbagefreeThreadContextMap";
+
+    private static final Lazy<LoggingSystem> SYSTEM = Lazy.relaxed(LoggingSystem::new);
+
+    private final Lock initializationLock = new ReentrantLock();
+    private volatile LoggingProvider provider;
+    private final Lazy<LoggerContextFactory> loggerContextFactorySupplier = Lazy.lazy(() -> getProvider().createLoggerContextFactory());
+    private Supplier<ThreadContextMap> threadContextMapSupplier = () -> getProvider().createContextMap();
+    private Supplier<ThreadContextStack> threadContextStackSupplier = () -> getProvider().createContextStack();
+
+    /**
+     * Acquires a lock on the initialization of locating a logging system provider. This lock should be
+     * {@linkplain #releaseInitializationLock() released} once the logging system provider is loaded. This lock is
+     * provided to allow for lazy initialization via frameworks like OSGi to wait for a provider to be installed
+     * before allowing initialization to continue.
+     *
+     * @see <a href="https://issues.apache.org/jira/browse/LOG4J2-373">LOG4J2-373</a>
+     */
+    public void acquireInitializationLock() {
```

Comment:
This class is not marked internal in any way. Exposing locks publicly like this is very dangerous.  It would make more sense to me to provide a method that accepts a Lambda to performed while the lock is held.

Reply:
Fixed. Thanks for catching that!